### PR TITLE
[SPARK-27539][SQL] Fix inaccurate aggregate outputRows estimation with column containing null values

### PR DESCRIPTION
**Pull request by pengbo**
## What changes were proposed in this pull request?
This PR is follow up of https://github.com/apache/spark/pull/24286. As @gatorsmile pointed out that column with null value is inaccurate as well.

```
> select key from test;
2
NULL
1
spark-sql> desc extended test key;
col_name key
data_type int
comment NULL
min 1
max 2
num_nulls 1
distinct_count 2
```

The distinct count should be distinct_count + 1 when column contains null value.
## How was this patch tested?

Existing tests & new UT added.

### CODE REVIEW

**Comment by pengbo** [comment]
test it please

**Comment by cloud-fan** [comment]
ok to test

**Review by srowen — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Merged to master/2.4.
cc @gatorsmile

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/AggregateEstimation.scala [hunk 1]
```diff
@@ -42,8 +42,8 @@ object AggregateEstimation {
         (res, expr) => {
           val columnStat = childStats.attributeStats(expr.asInstanceOf[Attribute])
           val distinctCount = columnStat.distinctCount.get
-          val distinctValue: BigInt = if (distinctCount == 0 && columnStat.nullCount.get > 0) {
-            1
+          val distinctValue: BigInt = if (columnStat.nullCount.get > 0) {
+            distinctCount + 1
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/AggregateEstimation.scala]: Hm, actually, do we need to count `null` as distinct value? It's not counted as a distinct value in SQL (`F.countDistinct` or `count(DISTINCT col)`) and Pandas (`unique()` by default) at least.

Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/AggregateEstimation.scala]: Looking into the current impl, seems we ignore `null` as distinct values:

https://github.com/apache/spark/blob/239082d9667a4fa4198bd9524d63c739df147e0e/sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala#L270
https://github.com/apache/spark/blob/b1857a4d7dfe17663f8adccd7825d890ae70d2a1/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala#L50-L51

**Comment by HyukjinKwon** [comment]
Reverting this. I don't think this is a correct fix.

**Comment by dongjoon-hyun** [comment]
I added comment on the original PR. The reverting looks wrong to me.
- https://github.com/apache/spark/pull/24286#issuecomment-495351431

**Pull request closed without merging**